File: sast-go/cases/accuracy/context_sensitive/polymorphism/polymorphism_002_F/polymorphism_002_F.go

// evaluation information start
// real case = false
// evaluation item = 准确度->上下文敏感分析->多态
// scene introduction = 继承
// level = 2 
// bind_url = accuracy/context_sensitive/polymorphism/polymorphism_002_F/polymorphism_002_F
// evaluation information end

package polymorphism_002_F

func polymorphism_002_F(__taint_src interface{}) {
	var sub Base
	sub = Sub2{}
	__taint_sink(sub.call(__taint_src))
}

type Base interface {
	call(src interface{}) interface{}
}

type Sub1 struct{}

func (s Sub1) call(src interface{}) interface{} {
	return src
}

type Sub2 struct{}

func (s Sub2) call(src interface{}) interface{} {
	return "_"
}

func __taint_sink(o interface{}) {
}

File: sast-php/case/accuracy/context_sensitive/polymorphism/polymorphism_002_F.php
<?php
// evaluation information start
// real case = false
// evaluation item = 准确度->上下文敏感->多态
// scene introduction = 抽象类多态，子类实现返回安全数据
// level = 3
// bind_url = accuracy/context_sensitive/polymorphism/polymorphism_002_F
// evaluation information end

function __taint_sink($o) {
    shell_exec($o);
}

abstract class BaseHandler {
    abstract public function handle($data);
}

class TaintHandler extends BaseHandler {
    public function handle($data) {
        return $data;
    }
}

class SafeHandler extends BaseHandler {
    public function handle($data) {
        return "safe";
    }
}

function polymorphism_002_F($__taint_src) {
    $handler = new SafeHandler();
    $result = $handler->handle($__taint_src);
    __taint_sink($result);
}

$taint_src = "taint_src_value";
polymorphism_002_F($taint_src);
